sycl/doc/extensions/proposed/sycl_ext_oneapi_annotated_ptr.asciidoc

```diff
@@ -0,0 +1,650 @@
+
+The purpose of the `annotated_ref` class template is to provide reference
+wrapper semantics. It enables properties to be preserved on loads from and
+stores to annotated_ptrs.
+
+```c++
+namespace sycl::ext::oneapi::experimental {
+template <typename T, typename PropertyListT = properties<>>
+class annotated_ref {
+  public:
+    annotated_ref(T *);
+    operator T() noexcept;
+    operator const T() const noexcept;
+    void operator=(const T &);
+  private:
+    T *Ptr;
+  };
+} // namespace sycl::ext::oneapi::experimental
+```
+
+
+Member Functions are described in the table below
+[frame="topbot",options="header"]
+|===
+|Functions |Description
+
+// --- ROW BREAK ---
+a|
+[source,c++]
+----
+annotated_ref(T * InputPtr);
+----
+|
+Constructs an `annotated_ref` object. Does not allocate new storage. The
+underlying pointer is initialized to `InputPtr`.
+
+// --- ROW BREAK ---
+a|
+[source,c++]
+----
+operator T() noexcept;
+operator const T() const noexcept;
+----
+|
+Reads the object of type `T` that is referenced by this wrapper, applying the
+annotations when the object is loaded from memory.
+
+// --- ROW BREAK ---
+a|
+[source,c++]
+----
+void operator=(const T &);
+----
+|
+Writes an object of type `T` to the location referenced by this wrapper,
+applying the annotations when the object is stored to memory.
+
+|===
+
+=== Properties
+
+Below is a list of compile-time constant properties supported with
+`annotated_ptr`. Other extensions can define additional compile-time constant
+properties that can be supported with `annotated_ptr`. Runtime properties
+are not supported.
+
+```c++
+namespace sycl::ext::oneapi::experimental {
+struct alignment_key {
+  template<int K>
+  using value_t = property_value<alignment_key, std::integral_constant<int, K>>;
+};
+
+struct runtime_aligned_key {
+  using value_t = property_value<runtime_aligned_key>;
+};
+
+template<int K>
+inline constexpr alignment_key::value_t<K> alignment;
+inline constexpr runtime_aligned_key::value_t runtime_aligned;
+
+template<>
+struct is_property_key<alignment_key> : std::true_type {};
+template<>
+struct is_property_key<runtime_aligned_key> : std::true_type {};
+
+template<typename T, typename PropertyListT>
+struct is_property_key_of<
+  alignment_key, annotated_ptr<T, PropertyListT>> : std::true_type {};
+template<typename T, typename PropertyListT>
+struct is_property_key_of<
+  runtime_aligned_key, annotated_ptr<T, PropertyListT>> : std::true_type {};
+} // namespace sycl::ext::oneapi::experimental
+```
+--
+[options="header"]
+|====
+| Property | Description
+|`alignment`
+|
+Indicates the alignment of the pointer in bytes. Alignment cannot be smaller
+than the size of type `T`.
+
+|`runtime_aligned`
+|
+Informs the compiler that the pointer has at least the default alignment for a
+pointer allocated through the SYCL runtime. This is always safe to apply to any
+pointer returned by a SYCL memory allocation function.
```

Comment:
Joe explained that keeping this looser wording enables more use cases, example:
```
auto ptr = malloc_shared(...)
auto ptr2 = ptr + (kAlignment * N);
```
In this case even `ptr2` can safely retain the `runtime_aligned` property. Note that this is legal only if the user offset `ptr1` by some multiple of the alignment value. If we renamed this property to something like "usm_returned" maybe then we can't do this.

Reply:
How does the application determine what the "default alignment" is? Unless there is some way to query this, I don't see how we can enable the use case you show.

In another conversation, we are considering adding the following extension:

* A query which returns the set of supported alignments for USM allocations.
* A runtime property that is accepted by the USM allocation functions, which allows the application to choose the alignment of the allocated memory.  The user would be responsible for choosing one of the supported alignments, otherwise the allocation request will fail.

What would "default alignment" mean if we do this?  If the application uses the property to request an alignment, is the pointer still "default aligned"?

I'm hesitant to define the `runtime_aligned` property unless / until we can define it precisely.

Reply:
Yes it seems we need the mechanism to query the alignment. We can add this property later when we have all pieces ready to support these use-cases.

Reply:
OK.  Since this is the only property in the `sycl_ext_oneapi_kernel_arg_properties` extension, we can remove that extension from this PR and add it later.

Reply:
No this `runtime_aligned` property is part of the core `annotated_ptr` extension. You are thinking of `restrict` which is the only property in `sycl_ext_oneapi_kernerl_arg_properties`

Reply:
Removed the property and opened an issue for it in the Issues section.

Reply:
> No this runtime_aligned property is part of the core annotated_ptr extension. You are thinking of restrict which is the only property in sycl_ext_oneapi_kernerl_arg_properties

Oh right, my mistake.  We can still add the property later.  We'll just update the feature-test version of the specification when we add it.